Top Pane and User Button

creviera opened the pull request:
## Description

Resolves https://github.com/freedomofpress/securedrop-client/issues/279
Resolves https://github.com/freedomofpress/securedrop-client/issues/298

## Summary of changes

#### UX Review needed

* [Behavior] Now, instead of having to click directly on the dropdown icon, a user can click anywhere on the `UserButton` to open the menu.

* [Behavior] Instead of hiding the refresh icon when "offline" disable it to show disabled image. 

* [Behavior] Added Static & Active states to the refresh icon, see freedomofpress/securedrop-ux#17 (comment)

* [Behavior] Before error messages were not displayed to the user. Now they show up in the center of the top pane for 10 seconds or until replaced with a new error message.

* [Styling] Added text styles for Refresh timestamp, journalist username, and sign-in button, see https://app.zeplin.io/project/5c807ea562f734bd2756b243/dashboard

* [Styling] Added a vertical bar and error icon next to the error status message.

#### Other changes

* Added background gradients on the top pane, left pane, and sign-in button as placeholders until final design is finished.

## Test Plan

  * Click on journalist name to see drop-down menu pop up
  * Click on drop-down arrow to see drop-down menu pop up
  * When signed out, click on star or delete icons to see 10-second error message centered in top pane
  * Test that error pop hides when a message is cleared because of logging in. To do this, hit the star or delete icons to trigger the error message, then log in before 10 seconds is up to see error pop up go away.
  * When signed in, click on refresh and see active-state refresh icon
  * Sign out and see refresh icon change to the disabled image

## A little demo

![demo3](https://user-images.githubusercontent.com/4522213/55387758-df797100-54e6-11e9-91a1-1869c0e4491e.gif)

creviera commented on securedrop_client/resources/__init__.py:
```diff
@@ -63,3 +63,32 @@ def load_css(name):
+    icon = QIcon()
+
+    pixmap = load_image(normal)
+    scaled = pixmap.scaled(18, 18, Qt.KeepAspectRatio, transformMode=Qt.SmoothTransformation)
```
You'll notice I'm using `Qt.SmoothTransformation` for smoothing which uses bilinear filtering under the hood. Some of the svgs will need to be replaced because opening them directly in the browser shows that they are small and pixel-y so they will not scale nicely. But for now, the smoothing effect makes it look much better than using `FastTransformation` or not scaling.

creviera replied:
Looks like the dropdown arrow svg is just rendered horribly in my firefox applicaiton on linux as well as mac, so the problem wasn't how svgs were created and exported in Sketch, but rather how we're loading them into Qt widgets. After making a change to load svgs into icons using `addFile` or labels using `QSvgWidget` they now should look :sparkles: crispy (not blocky)

creviera commented on securedrop_client/gui/main.py:
```diff
@@ -197,9 +192,22 @@ def show_conversation_for(self, source: Source, is_authenticated: bool):
         self.main_view.set_conversation(conversation_container)
 
-    def set_status(self, message, duration=0):
+    def update_activity_status(self, message: str, duration=0):
```
`update_activity_status` is a clearer name than `set_status` which I noticed was accidentally being used to update some error statuses in `logic.py`.

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -19,38 +19,107 @@
-            QPushButton { border: none; }
+            QPushButton { border: none; color: #fff }
+            QStatusBar#activity_status_bar { color: #fff; }
+            QWidget#error_bar { background-color: #f22b5d }
```
Note: I got the values of the colors using Gimp's color-sampling tool on images from here: https://projects.invisionapp.com/share/XDRB535NR9Z, which contains wireframes for the error status bar, etc (click through the story). I'm grabbing these colors because they are more up-to-date in invisionapp than they are in zeplin. So this is just to get it to look more like the design we're leaning towards. Eventually we'll code in a better way to store and use a color scheme.

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -64,31 +133,54 @@ def on_refresh_clicked(self):
         """
-        self.refresh.hide()
+        self.error_status_bar.showMessage(message, duration)
+        self.error_status_timer.start(duration)
```
The error status widget is timed to popup for the same amount of time as we display the message.

heartsucker commented:
Clicking sign out fails:

```pytb
Traceback (most recent call last):
  File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1282, in _show_or_hide_replybox
    self.reply_box.disable()
  File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1331, in disable
    self.text_edit.setText(_('You need to log in to send replies.'))
RuntimeError: wrapped C/C++ object of type QTextEdit has been deleted
```

heartsucker commented:
The reply box for typing and sending replies is not present for me.

![Screenshot_2019-04-10_14-23-12](https://user-images.githubusercontent.com/3998464/55878256-5aecb980-5b9c-11e9-9b18-1e4e584c9edf.png)



I know that Debian / xmonad is not the standard env we are targeting, but I think we will need to consider supporting Debian and macOS for developer agility and to allow outside contributions.

heartsucker commented:
Attempting to delete a source throws an exception:

```pytb
Traceback (most recent call last):
  File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/main.py", line 177, in on_source_changed
    self.show_conversation_for(self.current_source, self.controller.is_authenticated)
  File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/main.py", line 190, in show_conversation_for
    is_authenticated)
  File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1264, in __init__
    parent=self)
  File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1170, in __init__
    self.update_conversation(self.source.collection)
  File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1184, in update_conversation
    self.add_message(conversation_item)
  File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1218, in add_message
    self.controller.session.refresh(message)
  File "/home/heartsucker/.local/share/virtualenvs/securedrop-client-GExBtLKv/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 1498, in refresh
    self._expire_state(state, attribute_names)
  File "/home/heartsucker/.local/share/virtualenvs/securedrop-client-GExBtLKv/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 1600, in _expire_state
    self._validate_persistent(state)
  File "/home/heartsucker/.local/share/virtualenvs/securedrop-client-GExBtLKv/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 2047, in _validate_persistent
    state_str(state))
sqlalchemy.exc.InvalidRequestError: Instance '<Message at 0x7f2f28029eb8>' is not persistent within this Session
```

creviera commented:
> Clicking sign out fails:
> 
> ```
> Traceback (most recent call last):
>   File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1282, in _show_or_hide_replybox
>     self.reply_box.disable()
>   File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1331, in disable
>     self.text_edit.setText(_('You need to log in to send replies.'))
> RuntimeError: wrapped C/C++ object of type QTextEdit has been deleted
> ```

This has to do with how reply box was being deleted before. Since deleting the reply box is no longer necessary, removing that code fixed the bug.

creviera commented:
> The reply box for typing and sending replies is not present for me.

Can you try now that I removed the old code where we deleted the reply box widget?

creviera commented:
> Attempting to delete a source throws an exception:
> 
> ```
> Traceback (most recent call last):
>   File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/main.py", line 177, in on_source_changed
>     self.show_conversation_for(self.current_source, self.controller.is_authenticated)
>   File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/main.py", line 190, in show_conversation_for
>     is_authenticated)
>   File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1264, in __init__
>     parent=self)
>   File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1170, in __init__
>     self.update_conversation(self.source.collection)
>   File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1184, in update_conversation
>     self.add_message(conversation_item)
>   File "/home/heartsucker/code/freedomofpress/securedrop-client/securedrop_client/gui/widgets.py", line 1218, in add_message
>     self.controller.session.refresh(message)
>   File "/home/heartsucker/.local/share/virtualenvs/securedrop-client-GExBtLKv/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 1498, in refresh
>     self._expire_state(state, attribute_names)
>   File "/home/heartsucker/.local/share/virtualenvs/securedrop-client-GExBtLKv/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 1600, in _expire_state
>     self._validate_persistent(state)
>   File "/home/heartsucker/.local/share/virtualenvs/securedrop-client-GExBtLKv/lib/python3.5/site-packages/sqlalchemy/orm/session.py", line 2047, in _validate_persistent
>     state_str(state))
> sqlalchemy.exc.InvalidRequestError: Instance '<Message at 0x7f2f28029eb8>' is not persistent within this Session
> ```
> 
> This may be related to what I suggested in #296.

I created an issue for tracking this here: https://github.com/freedomofpress/securedrop-client/issues/297

eloquence commented:
Agreed upon to-dos:
- Summarize PR title, clarify description
- Break out reply behavior change
- Clarify test plan

ninavizz commented:
Clarification to "text styles': All text styles have font-family, font-weight, color, line-spacing, character-spacing, padding, and margin attributes. It's impossible to "verify/review" only one or two of those things, outside from the others (other than color, which can be done with an eyedropper tool). FYI. :)

creviera commented:
@ninavizz sounds good to me. let's verify all of them in the next UX Review meeting. I added

ninavizz commented:
Cool. I'll timebox some time for myself on Monday to do more organizing/syleguiding w/in Zeplin, too.

creviera commented:
@emkll this is now ready for your eyes :)

creviera commented:
> Clarify test plan

This is done

creviera commented:
> Agreed upon to-dos:
> 
>     * Summarize PR title, clarify description
> 
>     * Break out reply behavior change
> 
>     * Clarify test plan

These tasks are done.

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -19,181 +19,553 @@
+    A button that shows an icon for different refresh states.
+    """
+
+    css = '''
```
Nit: class constant should be `SCREAMING_SNAKE_CASE`

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -19,181 +19,553 @@
+    displayed for a given duration or until the message updated with a new message.
+    """
+
+    css = '''
```
Nit: class constant should be `SCREAMING_SNAKE_CASE`

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -19,181 +19,553 @@
+    be displayed for a given duration or until the message is cleared or updated with a new message.
+    """
+
+    css = '''
```
Nit: class constant should be `SCREAMING_SNAKE_CASE`

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -19,181 +19,553 @@
+        self.status_timer.timeout.connect(self._on_status_timeout)
+
+    def _hide(self):
+        self.vertical_bar.hide()
```
We should make sure each of these objects has the `parent` value set so that hiding the parent hides all the children. This would make the show/hide methods unnecessary. I don't know if this is possible in this context, and if not it might make sense for us to rewrite it work like that. I think in general in our code we should be using the parent/child relationships more, and this is one place it seems like it would be beneficial.

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -19,181 +19,553 @@
 
-    def __init__(self, parent: QWidget):
-        super().__init__(parent)
+    css = '''
```
Nit: class constant should be `SCREAMING_SNAKE_CASE`

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -19,181 +19,553 @@
+        self.user_icon.setText(_('jo'))
+        self.user_button.set_username(username)
+
+    def show(self):
```
This is another case where the show/hide methods could possibly be replaced by parent/child + show/hide of the parent.

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -19,181 +19,553 @@
+    This button is responsible for launching the journalist menu on click.
+    """
+
+    css = '''
```
Nit: class constant should be `SCREAMING_SNAKE_CASE`

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -19,181 +19,553 @@
+    """
 
-        self.user_menu.show()
+    css = '''
```
Nit: class constant should be `SCREAMING_SNAKE_CASE`

creviera replied:
lol good catch

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -215,16 +588,12 @@ def __init__(self, parent):
         self.view_layout = QVBoxLayout()
         self.view_layout.setContentsMargins(0, 0, 0, 0)
         self.view_holder = QWidget()
+        self.view_holder.setStyleSheet('background: #fff;')
```
We may want to come up with a standard practice of where we place CSS. Right now it seems some is in class constants and some is inline as strings. To help us find CSS later, we might want to make all of them class constants so we have a better idea where to look. This doesn't need to be fixed here, just thinking out loud.

emkll replied:
This is a good point, I think it will also help with maintainability and decoupling of style/functionality.

creviera replied:
So I decide to move all of our CSS to a class string var for each widget that we want to style. I also started labeling our widgets with object names to be used in CSS. This is a first step in moving out our CSS into CSS files.

creviera replied:
I might have missed some areas in the code, but will slowly be refactoring each widget in this way. Take a look at all of the new widget classes that I created for example.

heartsucker commented on securedrop_client/gui/widgets.py:
```diff
@@ -562,7 +937,7 @@ class SpeechBubble(QWidget):
     """
 
-    css = "padding:8px; min-height:32px; border:1px solid #999; border-radius:18px;"
+    css = "padding:8px; min-height:32px; border:1px solid #999;"
```
Nit: class constant should be `SCREAMING_SNAKE_CASE`

creviera commented:
Thanks for the great feedback @emkll! Here's my followup:
* 1 - https://github.com/freedomofpress/securedrop-client/issues/301
* 2 - The "delete source acount" popup is not in the scope of this PR, however this is great feedback. I moved your comment to an issue for working on the Source Menu.  Here I just focused on the User Menu, which still is only partially done, as you can see the menu itself needs to be styled.

creviera commented:
@heartsucker thanks for the final review. I think I addressed all of mickael's and your concerns. See all the Issues I created and linked to in this PR to keep the scope of this PR more narrow. Lemme know if I missed something!

I wonder if we should wait before merging any more PRs until the bugs on `master` are fixed? This PR shouldn't affect any of them like the SourceMenu segfault or segfault when deleting a Source (since this pr is related only to the TopPane and UserButton) but they didn't enter into the conversation during the review of this PR. And creating a Test Plan and wiki and all the Issues for things mentioned here outside of the scope of this PR took quite a bit of time.

If we're not confident about the client as is, let's spend some time to fix things. @eloquence thoughts?

eloquence commented:
How about this to get some sanity back:
- merge this PR if all outstanding issues specific to it are resolved
- prioritize segfaults/core functionality breakage (#302 and #297, any others)?
- bump most other styling/feature work off this sprint except maybe the small-ish #286

Would that make sense to you @creviera @redshiftzero ? It's only ~2 1/2 more days in the sprint so we're not talking about a huge delay.

creviera commented:
> I have some nits (basically just linting) and one place the code could be improved to use Qt best practices (parent/child relationships between widgets). None of these block and all of them can be followed up in other tickets. This PR makes some excellent changes, and I'd rather get them in than nit pick at it until it's "perfect." I'm going to leave it unmerged for now so @emkll can take a look too.

I created a place for us to discuss parent child relationships of our widgets, and also related, separation of concerns. For example, once a child has a parent, it can call methods on its parent, which makes it hard to reason about what each widget is responsible for. Often it looks like we pass parents to make it easier for children to tell the parent that they would like the parent, or its parent, or its parent, etc., to do a database operation. Anyway, I think this is a bigger discussion than how to design how we hide or show widgets. So go here to get opinionated: https://github.com/freedomofpress/securedrop-client/issues/303

creviera commented:
>     * prioritize segfaults/core functionality breakage (#302 and #297, any others)?

One other than I'm working on investigating and writing up a ticket for. I may have to open the issue without having an STR that consistently produces the segfault.

>     * bump most other styling/feature work off this sprint except maybe the small-ish #286

Makes sense to me. I like the idea of figuring digging into `logic.py` more (good cross-team learning opportunity).

redshiftzero commented:
yep let's merge this and then focus on resolving bugs
